Loading manifests from blueprints

maxdup opened the pull request:
Hi, as mentioned in issue #20, flask blueprints can have their own static folders which flask-static-digest doesn't currently support. The static behavior for apps and blueprints is polymorphic through the `Flask.Scaffold` class. By that logic, what flask-static-digest does at the app level can easily be done at the blueprint level.

I've modified `flask_static_digest/__init__.py` to load manifests from blueprint's static_folder as well.

The `init_app` usage for flask-static-digest is unchanged. Only caveat is you'll want to register your blueprints before you call upon flask-static-digest

```
from flask import Flask
from flask_static_digest import FlaskStaticDigest
from .bp_my_blueprint import bp_mybp

flask_static_digest = FlaskStaticDigest()
app = Flask(__name__,
            static_folder='static')
...
app.register_blueprint(bp_mybp)
...
flask_static_digest.init_app(app)
```

My changes should have no side effects for other users and are ready to be shipped.
However, my PR is somewhat incomplete. I've only added blueprint support in `init_app`. 

Just let me know if you welcome changes for Blueprint support.
I may look into adding blueprint support in the cli/digester. I'm also willing to add tests and update documentation accordingly.

Cheers, thanks for this neat little piece of flask ecosystem :)

nickjj commented:
Hi,

Thanks for the PR.

I didn't run the code yet but you have `self.scaffold = scaffold or app` in `class _ScaffoldStaticDigest(object)`. What happens if you have both an app and a blueprint? Will it only add the blueprint's static files?

maxdup commented:
Essentially, the goal is to `app.add_template_global(static_url_for)` for both the app and each of its blueprints (whenever manifests are present).

This is achieved through this loop
```python
scaffolds = [app, *app.blueprints.values()] # the app and each of its registered blueprints                                                                       
for s in scaffolds:
    scaffold = _ScaffoldStaticDigest(app, s)
        if scaffold.has_manifest:
            app.add_template_global(scaffold.static_url_for)
```
The app will have its manifest loaded, as well as each of the blueprints. Each static folder will be served on separate urls, according to the static_url_path specified in the app or blueprint.


PS:
I realize my optional parameter on _ScaffoldStaticDigest.__init__() is ambiguous and unnecessary
What I meant was 
```python
class _ScaffoldStaticDigest(object):
    def __init__(self, app, scaffold):
        self.app = app
        self.scaffold = scaffold
```
But even there I'm not entirely satisfied with the implementation. I'll probably add a revision

maxdup commented:
PS: I haven't looked too deep into it but the CI seems to have failed because of a version problem between Flask and Jinja. I would suggest updating to `Flask==2.2.3`

here
https://github.com/nickjj/flask-static-digest/actions/runs/4289750320/workflow#L26

and here
https://github.com/nickjj/flask-static-digest/blob/master/tests/example_app/requirements.txt#L1
...and also flask-static-digest/tests/example_app/requirements.txt

maxdup commented:
Okay I've really taken a dive into flask to make sure I got everything right. The current implementation should make more sense if you want to have a look. 

I've gotten the github workflow to work again. 

I've gotten the `Flask.static_folder` working alongside a few `Blueprint.static_folder` in my current project. I'll try to replicate my success in the `tests/example_app` when I have a bit more time.

nickjj commented:
Thanks, I need to go through everything in more detail but one quick thing:

Will this work with both Flask 1.1 and 2.X?

maxdup commented:
Alright, this was quite the rabbit hole but I'm really happy with the results :)

I've fleshed out the example and squashed out a few edge case in the process. I believe we're at feature parity with url_for which is quite nice. Any static folder you can invoke with `url_for` can now contain manifests.

I've added a few test asserts for the new functionality.

I've added blueprint support in the cli for both the 'compile' and 'clean' commands.

I've made sure to keep support for flask 1.x.x
(I've included a `requirements-flask1x.txt` alongside the example if you wish to check it out. The tests will need to stay flask 2.0 tho, because we are testing flask 2 specific features)

So yeah I'm done here. Let me know if there's anything else you'd like to see

maxdup commented:
Just going to quickly reiterate the functionality this PR introduces so it's all clear :)

## Previously

The only use case supported by flask-static-digest was the flask app's static folder.
`{{ static_url_for('static', filename='css/app.css') }}`

## New!

The static folders specified by blueprints are now supported.
`{{ static_url_for('my_blueprint.static', filename='css/app.css') }}`

But also! you can access relative static folders with '.static'. This will map contextually to the correct static folder.
`{{ static_url_for('.static', filename='css/app.css') }}`

## New! (with flask 2.0.0+)

Blueprint nesting is now also supported.
`{{ static_url_for('my_blueprint.child_blueprint.static', filename='css/app.css') }}`

nickjj commented:
Hi,

This is looking good thanks. Would you say it's ready to be reviewed?

maxdup commented:
Definitely :)
Let me know how you feel about the cli output, I didn't really take any time to update it but it's mostly fine still.

nickjj commented:
Great, thanks. I'll review it for sure. Hopefully within a week or so. It's definitely not going to rot. It will be merged in the end, perhaps with a few minor changes which I'll call out in the review.

nickjj commented on tests/example_app/example/page1/page1a/bp_page1a.py:
```diff
@@ -0,0 +1,11 @@
+from flask import Blueprint, render_template
+
+bp_page1a = Blueprint('page_1a', __name__,
```
Can you change all of the page, page1a, page1b and page 2 references to use more "real" names? I found it to be a little hard to follow with everything named like this.

I'd suggest something like:

- `page1 -> blog`
  - `page1a -> comments`
  - `page1b -> I'm not sure`
- `page2 -> about`

Feel free to pick something else for all of these as long as it's cohesive.

I noticed there's no tests for `page1b` and no digested static files after I ran the digest command, was this intentional?

maxdup replied:
[e98d391 updated example app](https://github.com/nickjj/flask-static-digest/pull/33/commits/e98d3917c1f0ef23ef450c86c11f5b4f4dde4350)

Ended up going:
```
page1 -> about
    page1a -> our-story
    page1b -> our-team
page2 -> contact
```
Hope that makes the site structure more intuitive :)

nickjj replied:
Thanks, it does.

Can you take it 1 step further and change `our-story` to `story` and `our-team` to `team`?

maxdup replied:
[3174c64 example app code styling](https://github.com/nickjj/flask-static-digest/pull/33/commits/3174c64846fad1973666d27628e9f6cbc4ea9a04)

nickjj commented on tests/example_app/example/page1/templates/page_1.html:
```diff
@@ -0,0 +1,25 @@
+
+  <h2>Currently on page 1</h2>
+
+  <!-- relative static ".static" -->
```
You can remove all of these HTML comments in this file and `page_2.html`.

maxdup replied:
[d06ab59 favoring double quotes and removing html comments](https://github.com/nickjj/flask-static-digest/pull/33/commits/d06ab598dfca84d46029872fa82baf6f77884cfa)

nickjj commented on tests/example_app/example/page1/templates/page_1.html:
```diff
@@ -0,0 +1,25 @@
+  <!-- specific static "page_2.static" -->
+  <p>{{ static_url_for('page_2.static', filename='app.css') }}</p>
+
+
```
Can you remove this extra new line?

maxdup replied:
[27947c4 improved spacing consistency in jinja files](https://github.com/nickjj/flask-static-digest/pull/33/commits/27947c47ef46de2dbfc486e75705926ae6729b41)

nickjj commented on tests/example_app/example/page1/templates/page_1.html:
```diff
@@ -0,0 +1,25 @@
+
+  <link type="text/css" rel="stylesheet"
+        href="{{ static_url_for('.static', filename='app.css') }}">
+
```
Can you remove this extra new line?

maxdup replied:
[27947c4 improved spacing consistency in jinja files](https://github.com/nickjj/flask-static-digest/pull/33/commits/27947c47ef46de2dbfc486e75705926ae6729b41)

nickjj commented on tests/example_app/example/page2/templates/page_2.html:
```diff
@@ -0,0 +1,18 @@
+
+  <link type="text/css" rel="stylesheet"
+        href="{{ static_url_for('.static', filename='app.css') }}">
+
```
Can you remove this extra new line?

maxdup replied:
[27947c4 improved spacing consistency in jinja files](https://github.com/nickjj/flask-static-digest/pull/33/commits/27947c47ef46de2dbfc486e75705926ae6729b41)

nickjj commented on tests/example_app/example/page1/templates/page_1.html:
```diff
@@ -0,0 +1,25 @@
+        href="{{ static_url_for('.static', filename='app.css') }}">
+
+{% endblock %}
+
```
Can you remove this extra new line?

maxdup replied:
[27947c4 improved spacing consistency in jinja files](https://github.com/nickjj/flask-static-digest/pull/33/commits/27947c47ef46de2dbfc486e75705926ae6729b41)

nickjj commented on tests/example_app/example/page1/templates/page_1.html:
```diff
@@ -0,0 +1,25 @@
+
+{% endblock %}
+
+
```
Can you remove this extra new line?

maxdup replied:
[27947c4 improved spacing consistency in jinja files](https://github.com/nickjj/flask-static-digest/pull/33/commits/27947c47ef46de2dbfc486e75705926ae6729b41)

nickjj commented on tests/example_app/example/page1/page1a/templates/page_1a.html:
```diff
@@ -0,0 +1,17 @@
+
+  <link type="text/css" rel="stylesheet"
+        href="{{ static_url_for('.static', filename='app.css') }}">
+
```
Can you remove this extra new line?

maxdup replied:
[27947c4 improved spacing consistency in jinja files](https://github.com/nickjj/flask-static-digest/pull/33/commits/27947c47ef46de2dbfc486e75705926ae6729b41)

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -3,7 +3,7 @@
 from urllib.parse import urljoin
 
-from flask import url_for as flask_url_for
+from flask import _request_ctx_stack, url_for as flask_url_for
```
Is there any way we can do what we want here without depending on a private function from Flask?

maxdup replied:
[5f3b61e replaced flask._request_ctx_stack.top.request with flask.request](https://github.com/nickjj/flask-static-digest/pull/33/commits/5f3b61e20618e8deb18ac66759094ae94b50cf78)

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -26,28 +26,29 @@ def init_app(self, app):
+        self.manifests = {}
 
-        self.manifest = {}
+        def ingestScaffold(endpoint, scaffold):
```
As for naming, where did the phrase "scaffold" come from?

maxdup replied:
A Flask app inherits Scaffold: https://github.com/pallets/flask/blob/main/src/flask/app.py#L105
A Flask blueprint inherits Scaffold: https://github.com/pallets/flask/blob/main/src/flask/blueprints.py#L117

The Scaffold class is responsible for the static folder polymorphism between apps and blueprints.
https://github.com/pallets/flask/blob/2f67e0fe4aca8a8d6998b3022150f1e6d6738214/src/flask/scaffold.py#L54

nickjj replied:
Thanks for the clarification. If that's how it's defined at the Flask level I'm ok with the name scaffold. I'll make a follow up comment on suggestions about the function's logic.

nickjj replied:
This comment applies to the following code:

```py
        def ingestScaffold(endpoint, scaffold):
            if not scaffold.has_static_folder:
                return

            manifest_path = os.path.join(scaffold.static_folder,
                                         "cache_manifest.json")
            has_manifest = os.path.exists(manifest_path)

            if has_manifest:
                manifest = self._load_manifest(scaffold, manifest_path)
                self.manifests[endpoint] = manifest

        ingestScaffold("static", app)
        for endpoint, scaffold in app.blueprints.items():
            ingestScaffold(f"{endpoint}.static", scaffold)
```

Can we:

- Rename `ingestScaffold` to `load_manifests`

Also are we going to get into trouble by hard coding `("static", app)` and `f"{endpoint}.static"`. Are these always guaranteed to be named `static` no matter what type of customization an end user will do using Flask's public interface?

nickjj replied:
Also what do you think about renaming `for endpoint, scaffold in app.blueprints.items():` to `for endpoint, blueprint in app.blueprints.items():`. I think it's a little more clear since we're looping over blueprint items. The blueprint is the item in the loop. That would mean changing the line below it to reference `blueprint` too.

maxdup replied:
I'm in favor of renaming `ingestScaffold` to `load_manifest`

I'll begrudgingly rename the `scaffold` argument to `blueprint`.
It's just that `def load_manifest(endpoint, blueprint):` is going to feel awkward when the argument isn't a blueprint.
i.e. `load_manifest("static", app)`.

Static endpoints are hardcoded as "static" here:
https://github.com/pallets/flask/blob/main/src/flask/app.py#L513
https://github.com/pallets/flask/blob/main/src/flask/blueprints.py#L322

I'm for sure going to review how blueprints are iterated, along with the CLI

nickjj replied:
Ok thanks, sounds good.

Let's roll with:

- `load_manifest` and keep its argument named as `scaffold`
- Keep the hard coded `static`

nickjj replied:
Where do you stand on https://github.com/nickjj/flask-static-digest/pull/33#discussion_r1133085554? This one feels like it would always be a blueprint, in which case we could change its name?

maxdup replied:
[0501677 renaming scaffold and ingestScaffold, reworked load_manifest](https://github.com/nickjj/flask-static-digest/pull/33/commits/05016779a5593dd065231c662fd6040c2fb6e09e)

Turns out there was already a function named `_load_manifest`. I ended up merging `ingestScaffold` with `_load_manifest` and reworked the code a little bit as I went. The file IO is going to be more robust now and less likely to crash the server.

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -26,28 +26,29 @@ def init_app(self, app):
-            manifest_dict = json.load(f)
+        ingestScaffold('static', app)
+        for endpoint, scaffold in app.blueprints.items():
+            ingestScaffold(endpoint + '.static', scaffold)
```
Can you use an f-string here instead of `+`?

maxdup replied:
[7a32077 using fstrings, updated CI](https://github.com/nickjj/flask-static-digest/pull/33/commits/7a32077878a508fa9cb0aae820a121ba32875a62)

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -26,28 +26,29 @@ def init_app(self, app):
-    def _load_manifest(self, app):
-        with app.open_resource(self.manifest_path, "r") as f:
-            manifest_dict = json.load(f)
+        ingestScaffold('static', app)
```
This project is using double quotes instead of single. Can you use double quotes everywhere?

maxdup replied:
[d06ab59 favoring double quotes and removing html comments](https://github.com/nickjj/flask-static-digest/pull/33/commits/d06ab598dfca84d46029872fa82baf6f77884cfa)

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -26,28 +26,29 @@ def init_app(self, app):
-        return urljoin(self.host_url,
-                       "/".join([self.static_url_path, filename]))
+    def _prepend_host_url(self, flask_url):
+        return urljoin(self.host_url, flask_url)
```
Was it intended to change this from its original behavior?

maxdup replied:
Respectfully, the previous `"/".join([self.static_url_path, filename]` was a bad approach to creating urls which I ditched entirely. 
It reimplements the functionality of flask's url_for and wasn't going support routing to blueprints moving forward. I've removed it in favor of always relying on flask's `url_for`. very much intentional.

More thoughts on that in the next conversation item.

nickjj replied:
I vaguely remember there being a problem when the `host_url` was more than a domain such as `cdn.example.com`. For example if you had `cdn.example.com/custom/path` I think `urljoin` didn't work out of the box. Or maybe it was a problem when the `static_url_path` had nested URL paths. It was like 4 years ago.

The general problem was making `urljoin` work with multiple parts. I'm pretty sure I got that implementation from https://stackoverflow.com/questions/36189261/python-join-multiple-components-to-build-a-url.

maxdup replied:
Well the good news is you were already joining paths using urljoin so no regression here.

If flask's `url_for` can't produce the correct url, someone is undoubtedly misusing flask and we shouldn't be expected to support that. I don't care if it breaks for people if it's in line with Flask's expected behavior.

nickjj replied:
I think originally I had the extra `join` in there so the extra parts were handling by the join and then `urljoin` only has to deal with 2 parts since the `static_url_path` and filename parts were handled by the `join`.

For example, here's what happens when you don't do the extra `join` and you have more than 2 parts:

```py
>>> urljoin("https://cdn.example.com", "hello", "app.css")
'https://cdn.example.com/hello'
```

`urljoin` by itself will not support 3 arguments in the way we want.

But my original approach does:

```py
>>> urljoin("https://cdn.example.com", "/".join(["hello", "app.css"]))
'https://cdn.example.com/hello/app.css'
```

It also supports the extra slash if it happens to exist:

```py
>>> urljoin("https://cdn.example.com", "/".join(["/hello", "app.css"]))
'https://cdn.example.com/hello/app.css'
```

But I guess with your new implementation, `flask_url` already handles joining the `static_url_path` and the asset's file name so `urljoin` is only ever dealing with 2 parts so it works? I dropped in a couple of prints and see `flask_url` is equal to `/static/css/app-d41d8cd98f00b204e9800998ecf8427e.css`.

I think we're good with leaving the new version.

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -61,26 +62,40 @@ def static_url_for(self, endpoint, **values):
-            return flask_url_for(endpoint, **merged_values)
+        flask_url = flask_url_for(endpoint, **merged_values)
+
+        return self._prepend_host_url(flask_url)
```
Previously we returned 2 different things based on if `if self.host_url` was set when folks wanted to prepend an optional CDN URL. Does this break that?

maxdup replied:
Note: [4a5827e](https://github.com/nickjj/flask-static-digest/pull/33/commits/4a5827e6fc984a9230da71291e4e78629baaf008) I've just removed to `_prepend_host_url` to make this even simpler and hopefully clearer.

So the old implementation of `static_url_for` had 4 different returns
``` python
if not self.has_manifest:
    if self.host_url:
        return ....
    else:
        return ...

if self.host_url:
   return ...
else:
   return ...
```

Yes, at first glance, there's way less logic branches in my implementation but that doesn't mean less cases are being handled. Hear me out.

``` python
# We have a set of values we're going to supply to url_for.
# At this point, 'new_filename' may or may not have a replaced filename
# The 'if self.has_manifest' case was already handled beforehand
merged_values = {**values, **new_filename}

# flask_url_for will give us a relative path which we always want to have as a starting point
flask_url = flask_url_for(endpoint, **merged_values)

# Here, self.host_url may or may not be None.
# if 'self.host_url' happens to be None, urljoin will simply return 'flask_url', unchanged
# The 'if self.host_url' case is already built into 'urljoin'
return urljoin(self.host_url, flask_url)
```

Main point here, the 4 different returns weren't so different after all. They don't really change our course of action.
In every case, we want to `url_for` with the appropriate values.
I every case, it's safe to simply call urljoin and have it figure things out for us.

To answer your question specifically,  `self.host_url` is not broken and is still working as intended.


And here, the 'urljoin' implementation if you're curious:
https://github.com/python/cpython/blob/3.11/Lib/urllib/parse.py#L513

nickjj replied:
Does this still apply with my other comment of https://github.com/nickjj/flask-static-digest/pull/33#discussion_r1133083834?

nickjj replied:
Also in the original implementation this block of code existed at the top:

```py
        if not self.has_manifest:
            if self.host_url:
                return self._prepend_host_url(self.host_url,
                                              values.get("filename"))
            else:
                return flask_url_for(endpoint, **values)
```

This allowed us to exit early from the function when there's no digested files (no manifest).

I think this might introduce a performance regression because quite a bit more logic needs to be executed. Technically if there's no manifest then there's nothing to do since we can use Flask's `url_for` directly.

maxdup replied:
Having less logic branches leads to better performance. The CPU wants to speed through things. The more predictable your line of execution is, the faster it will go. I'd suggest reading about branchless programming and its benefits.

Besides, your early return optimizes for people who are wrongly using `static_url_for` in contexts where there are no manifests, it slows down every other correct use of `static_url_for`.
Frankly your concerns for performance are misguided. There's no doubt in my mind that performance will be impacted, but it's going to be from the added functionality, not a missing early return.

nickjj replied:
> Besides, your early return optimizes for people who are wrongly using static_url_for in contexts where there are no manifests, it slows down every other correct use of static_url_for

It's expected to _always_ use `static_url_for` when you're referencing any type of static file (css, js, images, etc.). The reason there is you can use it in development without digests or in production with (or without) digests without having to switch around your template tags for both cases.

> Frankly your concerns for performance are misguided. There's no doubt in my mind that performance will be impacted

Would you be open to benchmarking the original function and yours with and without digested files?

I'll admit I have not read the inner workings of how a CPU will process code at that level but I don't understand how executing all of the code below from your patch will end up being faster or equal vs the original function that exits early when there's no work to do by directly calling Flask's `url_for` and returning.

I do think when there's no digest the new version will be measurably slower. I really have no idea how much slower, it's possible it won't be impactful, but if we have the numbers in front of us with a repeatable test when we can see.

You may have multiple calls to `static_url_for` on a page (css + js + couple of images). Let's call it 5 calls on a page.

If your version takes on average 500 microseconds more than the original then we're talking 2.5ms. If your end to end server response time is 25ms, that's a 10% performance regression to render a response.

maxdup replied:
Well if I'm going to take the time to bench it, I'm also going to take the time to optimize it and that's probably a full rewrite of `static_url_for` incoming, as a fair warning.

But just to be clear, my implementation WILL BE SLOWER, that's the cost of new features. I'll be happy if I can make it to parity, if not then no biggie.

nickjj replied:
Maybe we can use Python's `timeit` module to get an idea on both implementations before rewriting?

If we determine it's only 50 microseconds slower then I don't think we need to rewrite it because a 1% regression for making the code more straight forward is worth it. Especially if that 25ms needs to travel 50ms over the internet and then there's 300ms spent painting a page in a browser. Suddenly it's a 375ms vs 250 microsecond difference in which case it's not worth the effort rewriting it.

maxdup replied:
Yeah it's a 25 microseconds increase for me, from 13 to 37 or so. Profiling reveals that most of that time is spent resolving endpoints which are now dynamic and `url_for` being slower than what `_prepend_host_url` was doing.

nickjj replied:
Would you mind sharing the code and commands you ran to profile it? I believe you but it would be nice to have a repeatable test. Maybe we can even commit it somewhere for future regression testing (I can do this work later if you want).

maxdup replied:
Just ran through this essentially, gives an average over 10000 calls
```
def create_app():
    app = Flask(__name__)

    app.config.from_object("config.settings")

    static_digest.init_app(app)

    @app.route("/")
    def index():

        start_time = timeit.default_timer()
        ITERS = 10000
        for i in range(ITERS):
            static_digest.static_url_for('static', filename='css/app.css')
            static_digest.static_url_for('static', filename='js/app.css')

        print((timeit.default_timer() - start_time) / ITERS)

        return render_template("index.html")

    return app
```
Kinda messy to set up since you'd need 2 virtual envs to compare. An arbitrary threshold for regressions could be set, but results will vary greatly on different hardware. You'd need an actual test suite powered by tox or something, could always start there.

nickjj replied:
Ok thanks. Let's keep your version.

Beyond that, there's a couple of change requests on renaming a couple of things and we're good to go to merge.

maxdup replied:
I had removed my benchmark so I redid it this morning and got much faster results somehow
old implementation:  7.55μs
new implementation: 9.8μs
My benchmark must have had some overhead or something, couldn't tell ya.

Simplest way to go about it just something like this:
```python
def create_app():
    app = Flask(__name__)

    app.config.from_object("config.settings")

    app.register_blueprint(bp_about)
    app.register_blueprint(bp_contact)

    flask_static_digest.init_app(app)

    @app.route("/")
    def index():
        ITERS = 10000
        time = timeit.timeit("flask_static_digest.static_url_for('static', " +
                             "filename='css/app.css')",
                             globals=globals(),
                             number=ITERS)
        print(time/ITERS)

        return render_template("index.html")

    return app
``` 
Not really something the current CI would support. You'd need an actual test suite. Probably with tox on top to manage multiple virtualEnvs if you're going to compare. Anyway the margin for error in benchmarking is rather large and I don't recommend introducing luck elements to your CI, that's generally a bad time waiting to happen.

nickjj replied:
Yep, I wouldn't block on CI based on results. I don't think this project really needs it, but if it did there would be benchmarks you could manually run on demand.

Thanks for setting all of this up.

nickjj commented on flask_static_digest/cli.py:
```diff
@@ -17,16 +17,27 @@ def digest():
-             current_app.config.get("FLASK_STATIC_DIGEST_BLACKLIST_FILTER"),
-             current_app.config.get("FLASK_STATIC_DIGEST_GZIP_FILES"))
+    for scaffold in [current_app, *current_app.blueprints.values()]:
+
```
Can you remove this extra new line?

maxdup replied:
[af8b1e3 fixed spacing in cli, removed unused imports](https://github.com/nickjj/flask-static-digest/pull/33/commits/af8b1e3de170227097090ef08ea0bb0a875d5b31)

nickjj replied:
Given the other proposed changes would it make sense to rename `for scaffold in` to `for blueprint in`?

maxdup replied:
[0501677 renaming scaffold and ingestScaffold, reworked load_manifest](https://github.com/nickjj/flask-static-digest/pull/33/commits/05016779a5593dd065231c662fd6040c2fb6e09e)

nickjj commented on flask_static_digest/cli.py:
```diff
@@ -17,16 +17,27 @@ def digest():
-           current_app.config.get("FLASK_STATIC_DIGEST_BLACKLIST_FILTER"),
-           current_app.config.get("FLASK_STATIC_DIGEST_GZIP_FILES"))
+    for scaffold in [current_app, *current_app.blueprints.values()]:
+
```
Can you remove this extra new line?

maxdup replied:
[af8b1e3 fixed spacing in cli, removed unused imports](https://github.com/nickjj/flask-static-digest/pull/33/commits/af8b1e3de170227097090ef08ea0bb0a875d5b31)

nickjj replied:
Given the other proposed changes would it make sense to rename `for scaffold in` to `for blueprint in`?

maxdup replied:
[0501677 renaming scaffold and ingestScaffold, reworked load_manifest](https://github.com/nickjj/flask-static-digest/pull/33/commits/05016779a5593dd065231c662fd6040c2fb6e09e)

nickjj reviewed (changes requested):
Hi,

I've dropped in a few change suggestions in a phase 1 review. Thanks for taking this on!

maxdup commented on tests/example_app/requirements-flask1x.txt:
```diff
@@ -0,0 +1,3 @@
+Flask==1.1.4
```
[03196f5 (example app) removed Flask 1.x and cleaned up imports](https://github.com/nickjj/flask-static-digest/pull/33/commits/03196f53e62fd9ce8561ef80819a155e6c4d4bac)

maxdup commented on tests/example_app/example/page1/bp_page1.py:
```diff
@@ -0,0 +1,25 @@
+                     url_prefix='/page-1',
+                     static_folder='static')
+
+# Just checking if flask version supports nested blueprints
```
[03196f5 (example app) removed Flask 1.x and cleaned up imports](https://github.com/nickjj/flask-static-digest/pull/33/commits/03196f53e62fd9ce8561ef80819a155e6c4d4bac)

maxdup commented on tests/example_app/example/app.py:
```diff
@@ -1,6 +1,9 @@
 from flask import Flask, render_template
 from flask_static_digest import FlaskStaticDigest
 
+from .page1.bp_page1 import bp_page1
```
[03196f5 (example app) removed Flask 1.x and cleaned up imports](https://github.com/nickjj/flask-static-digest/pull/33/commits/03196f53e62fd9ce8561ef80819a155e6c4d4bac)

maxdup commented:
I believe that covers it :)

I've closed the uncontroversial/straightforward things where I can respond with the appropriate changes.
I've left open the more open-ended things that might have to be discussed further and didn't yet result in any changes.

Ready for round 2, Let me know if there's anything else

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -61,26 +59,38 @@ def static_url_for(self, endpoint, **values):
+        flask_url = flask_url_for(endpoint, **merged_values)
+        return urljoin(self.host_url, flask_url)
+
+    def _load_manifest(self, scaffold, manifest_path):
```
Given the other proposed changes can you change the `scaffold` argument to `blueprint` here?

maxdup replied:
[0501677 renaming scaffold and ingestScaffold, reworked load_manifest](https://github.com/nickjj/flask-static-digest/pull/33/commits/05016779a5593dd065231c662fd6040c2fb6e09e)

nickjj commented on tests/example_app/example/about/our_story/bp_story.py:
```diff
@@ -0,0 +1,9 @@
+from flask import Blueprint, render_template
+
+bp_aboutStory = Blueprint("our_story", __name__,
```
Can you rename `bp_about_story` to `about_story`?

nickjj commented on tests/example_app/example/about/our_story/bp_story.py:
```diff
@@ -0,0 +1,9 @@
+
+
+@bp_aboutStory.route("/our-story")
+def pageAboutStory():
```
Can you rename this to `def story` with a URL of `/story` and also rename the HTML file to `story.html`?

nickjj commented on tests/example_app/example/about/our_story/templates/our_story.html:
```diff
@@ -0,0 +1,19 @@
+{% block body %}
+  <br/>
+  <br/>
+  <a href="/about/our-story">Our Story</a>
```
Can you make all of the rename suggestions from a previous comment and also use `url_for` instead of hard coding the URL?

nickjj commented on tests/example_app/example/about/our_team/bp_team.py:
```diff
@@ -0,0 +1,11 @@
+from flask import Blueprint, render_template
+
+bp_aboutTeam = Blueprint("our_team", __name__,
```
Can you make similar changes to the previous comment except using the word `team` instead of `story`?

nickjj commented on tests/example_app/example/about/our_team/templates/our_team.html:
```diff
@@ -0,0 +1,17 @@
+{% block body %}
+  <br/>
+  <br/>
+  <a href="/about/our-story">Our Story</a>
```
Can you make all of the rename suggestions from a previous comment and also use url_for instead of hard coding the URL?

nickjj commented on tests/example_app/example/about/our_story/templates/our_story.html:
```diff
@@ -0,0 +1,19 @@
+{% extends "layout.html" %}
+
+{% block body %}
+  <br/>
```
Can you remove both `<br/>` tags?

nickjj commented on tests/example_app/example/about/our_team/templates/our_team.html:
```diff
@@ -0,0 +1,17 @@
+{% extends "layout.html" %}
+
+{% block body %}
+  <br/>
```
Can you remove both `<br/>` tags?

nickjj commented on tests/example_app/example/about/templates/about.html:
```diff
@@ -0,0 +1,18 @@
+{% extends "layout.html" %}
+
+{% block body %}
+  <br/>
```
Can you remove both `<br/>` tags?

nickjj commented on tests/example_app/example/about/templates/about.html:
```diff
@@ -0,0 +1,18 @@
+{% block body %}
+  <br/>
+  <br/>
+  <a href="/about/our-story">Our Story</a>
```
Can you make all of the rename suggestions from a previous comment and also use url_for instead of hard coding the URL?

nickjj commented on tests/example_app/example/about/our_story/static/app.css:
```diff
@@ -0,0 +1,3 @@
+p {
+    color: blue;
```
Can you use 2 spaces instead of 4 and set `color: #0000ff;`?

nickjj commented on tests/example_app/example/about/static/app.css:
```diff
@@ -0,0 +1,3 @@
+p {
+    color: red;
```
Can you use 2 spaces instead of 4 and set `color: #ff0000;`?

nickjj commented on tests/example_app/example/about/our_team/static/app.css:
```diff
@@ -0,0 +1,3 @@
+p {
+    color: purple;
```
Can you use 2 spaces instead of 4 and set `color: #00ff00;`?

I have nothing against purple but the other colors were red and blue. Using green here completes the RGB trilogy.

nickjj commented on tests/example_app/example/contact/static/app.css:
```diff
@@ -0,0 +1,3 @@
+p {
+    color: green;
```
Can you use 2 spaces instead of 4 and set `color: #00ff00;`?

maxdup commented:
Alright so yeah I reworked a few things that were outside the scope of the feedback (mostly to do with `load_manifest`) but I'm done messing around with the code for real now. 
Let me know if you have any final notes or concern

nickjj commented on tests/example_app/example/about/story/bp_story.py:
```diff
@@ -0,0 +1,10 @@
+
+
+@bp_story.route("/story")
+def story():
```
Can you rename this from `story` to `index`?

nickjj commented on tests/example_app/example/about/story/bp_story.py:
```diff
@@ -0,0 +1,10 @@
+
+@bp_story.route("/story")
+def story():
+    return render_template("story.html")
```
Can you rename this from `story.html` to `index.html`?

nickjj commented on tests/example_app/example/about/story/templates/story.html:
```diff
@@ -1,19 +1,17 @@
-  <br/>
-  <a href="/about/our-story">Our Story</a>
-  <a href="/about/our-team">Our Team</a>
+  <a href="{{ url_for('about.story.story') }}">Our Story</a>
```
Can you rename the bp path to `about.story.index`?

nickjj commented on tests/example_app/example/about/story/templates/story.html:
```diff
@@ -1,19 +1,17 @@
-  <a href="/about/our-story">Our Story</a>
-  <a href="/about/our-team">Our Team</a>
+  <a href="{{ url_for('about.story.story') }}">Our Story</a>
+  <a href="{{ url_for('about.team.team') }}">Our Team</a>
```
Can you rename the bp path to `about.team.index`?

nickjj commented on tests/example_app/example/about/team/bp_team.py:
```diff
@@ -0,0 +1,11 @@
+
+
+@bp_team.route("/team")
+def team():
```
Can you rename this from `team` to index?

nickjj commented on tests/example_app/example/about/team/bp_team.py:
```diff
@@ -0,0 +1,11 @@
+
+@bp_team.route("/team")
+def team():
+    return render_template("team.html")
```
Can you rename this from `team.html` to `index.html`?

nickjj commented on tests/example_app/example/about/team/templates/team.html:
```diff
@@ -1,16 +1,14 @@
-  <br/>
-  <a href="/about/our-story">Our Story</a>
-  <a href="/about/our-team">Our Team</a>
+  <a href="{{ url_for('about.story.story') }}">Our Story</a>
```
Can you rename both bp paths to reference the new `.index` rename?

nickjj commented on tests/example_app/example/about/templates/about.html:
```diff
@@ -1,10 +1,8 @@
-  <br/>
-  <a href="/about/our-story">Our Story</a>
-  <a href="/about/our-team">Our Team</a>
+  <a href="{{ url_for('about.story.story') }}">Our Story</a>
```
Can you rename both bp paths (story and team) to reference the new `.index` rename?

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -26,28 +27,30 @@ def init_app(self, app):
+
+        manifest_path = os.path.join(scaffold._static_folder,
+                                     "cache_manifest.json")
+        try:
```
Nice touch on the error detection, since this only loads once when the extension loads it's worth it.

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -61,26 +64,22 @@ def static_url_for(self, endpoint, **values):
-        else:
-            return flask_url_for(endpoint, **merged_values)
+
+        # Note: This is taken from Flask's url_for
```
Can you remove these comments?

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -61,26 +64,22 @@ def static_url_for(self, endpoint, **values):
+        if request is not None:
+            blueprint_name = request.blueprint
+
+            # If the endpoint starts with "." and the request matches a
```
Can you remove these comments?

nickjj commented on flask_static_digest/__init__.py:
```diff
@@ -61,26 +64,22 @@ def static_url_for(self, endpoint, **values):
+        manifest = self.manifests.get(endpoint, {})
+        filename = values.get("filename", None)
+        values['filename'] = manifest.get(filename, filename)
+        return urljoin(self.host_url, flask_url_for(endpoint, **values))
```
Can you add 1 empty line above the `return` line?

maxdup commented:
Look brother. You can't have every template be named `index.html`. The app's template folder will take priority over all the other `index.html`.

I'm commited to this PR but god damn dude, god damn

nickjj commented:
> You can't have every template be named index.html.

I've done this in the past and it works out.

You could have a `base.html` layout and then:

You can have let's say (3) blueprints: `about`, `blog` and `contact`. It's not unreasonable to expect that each of those blueprints have an index page which would be the page associated with the following URLs: `/about/`, `/blog/` and `/contact/`. This makes the naming of them predictable across a project and avoids the "double" blueprint name reference such as `about.story.story`.

> I'm commited to this PR but god damn dude, god damn

I know, and I appreciate it but this is an open source project that had opinions prior to you committing to it. The example app is meant to act as documenting for anyone looking to use the extension.

A number of these opinion (such as double vs single quotes) could have been handled by using Black which I am going to add to this project after this PR is merged but some of them are stylistic changes based on how I've been personally building and organizing Flask apps for like 8 years.

maxdup commented:
Hey I'm not here to argue brother.
Coding style preferences are fine and I appreciate your thoroughness to be honest.

I've renamed templates to `index.html`. Can you help me understand why the CI is failing now?
(yes I'm being facetious, truly sorry dude I legit don't know how else to handle this)
(I was honest about thoroughness tho I swear)

nickjj commented:
It's ok. The commit you made looks good from scanning it.

I ran it locally and it looks like the digest changed so CI will need to be updated:

- `curl localhost:5000/about/` produces a 200 with a normal looking HTML response
- The HTML response has `app-d41d8cd98f00b204e9800998ecf8427e.css`
- The CI test that failed is grepping for `app-d41d8cd98f00b204e9800998ecf8427e.css`

Both the JS and CSS have the same digest in the HTML response on that page tho which looks like that might be unexpected?

`example/about/story/templates/index.html` is also referencing:

```html
  <p>{{ static_url_for("about.static", filename="app.css") }}</p>
  <p>{{ static_url_for("about.team.static", filename="app.css") }}</p>

  <link type="text/css" rel="stylesheet"
        href="{{ static_url_for('about.static', filename='app.css') }}">
  <link type="text/css" rel="stylesheet"
        href="{{ static_url_for('about.team.static', filename='app.css') }}">
```

Was it expected that you used `about.static` for both those, or should that be `about.story.static`?

maxdup commented:
> curl localhost:5000/about/ produces a 200 with a normal looking HTML response

Very true but unfortunately it's rendering the wrong `index.html`.

That's where you're getting the disparity between `app-d41d8cd98f00b204e9800998ecf8427e.css` and `app-297c671953f6c24f1354214b3f706e0b.css`

I'm using various `static`, `.static`, `about.static`, 'about.team.static' just to create diverse test cases. Validating that blueprints can use resources from each other and whatnot.

nickjj commented:
What if we rename that `example_app/templates/index.html` to `base.html`?

maxdup commented:
Hey technically that works. That's all I'm gonna say.

nickjj commented:
Cool. I typically roll with layout template names like:

- `base.html` (shared with all layouts)
- `app.html` (the main app)
- `account.html` (maybe a separate layout for login, account creation, etc.)
- `admin.html` (for an admin panel)

Using `index.html` never came up so I never ran into this issue where they compete.

maxdup commented:
Well idk, you've been using `layout.html` not `base.html`

Are you looking to name `example_app/templates/index.html` to `app.html` instead? Cause it's not gonna serve as a base.

nickjj commented:
> Are you looking to name example_app/templates/index.html to app.html instead?

Yeah that'll work.

maxdup commented:
Alright I believe this is it.

Wasn't prepared to wrap my head around `/` not being `index.html`. That's standard practice as old as the web.
Sorry about the confusion there.

nickjj commented on tests/example_app/example/about/story/templates/index.html:
```diff
@@ -0,0 +1,17 @@
+  <a href="{{ url_for('about.team.index') }}">Our Team</a>
+
+  <h2>Our Story</h2>
+  This is an example application
```
Can you change this to `<p>This is an example application.</p>` (with the period).

nickjj commented on tests/example_app/example/about/team/templates/index.html:
```diff
@@ -0,0 +1,15 @@
+  <a href="{{ url_for('about.team.team') }}">Our Team</a>
+
+  <h2>Our Team</h2>
+  <p>This is an example application</p>
```
Can you add a period at the end of the sentence?

nickjj commented on tests/example_app/example/about/templates/index.html:
```diff
@@ -0,0 +1,16 @@
+  <a href="{{ url_for('about.team.index') }}">Our Team</a>
+
+  <h2>About Us</h2>
+  <p>This is an example application</p>
```
Can you add a period at the end of the sentence?

nickjj commented on tests/example_app/example/contact/templates/index.html:
```diff
@@ -0,0 +1,13 @@
+
+{% block body %}
+  <h2>Contact Us</h2>
+  <p>This is an example application</p>
```
Can you add a period at the end of the sentence?

nickjj commented on tests/example_app/example/templates/layout.html:
```diff
@@ -9,6 +9,10 @@
 
 <body>
-  <h1>{% block body %}{% endblock %}</h1>
+  <h1>This is an example application</h1>
```
Can you add a period at the end of the sentence and also add 1 line break?

nickjj commented on tests/example_app/example/templates/app.html:
```diff
@@ -0,0 +1,5 @@
+{% extends "layout.html" %}
+
+{% block body %}
+  <p>Currently on page index</p>
```
Can you add a period at the end of the sentence?

nickjj commented:
Almost there, minor things left!

maxdup commented:
So look... I have no interest in maintaining test data that frankly should be discarded. I've only extended the favor so far because that's the lousy standard of proof that was set for this project.
You can't keep finding new ways to waste my time. I have warned you, yet you chose to move the goal post once again. Of all things, by requesting me to fix preexisting issues that are outside the scope of my PR.

With no end in sight, I find myself having to abandon this PR. Do with it as you please.

I promise I've tried but I unfortunately don't see any way for us to work together.
Your approach to code quality has been nothing but pedantic. It has lead you to wasting my time which could have been used far more effectively.
I cannot sit here and entertain your illusion of having high standards for code quality
(Admittedly, this is my own limitation as an equally flawed person, no hard feelings or anything).

Few points that stood out for me:

- We've found out that the linter in place is ineffective in producing the PR you're looking for.
- The test suite is an inadequate bash script.
- The code coverage is inexistent.
- There is no platform validation (MacOS/Windows)
- The documentation is dense and long-winded which makes it difficult to extract useful information.

These points undermine flask-static-digest in its ability to shine as a quality extension (More so than any PR could).

This could all have been fixed. I was excited to bring new functionality to flask-static-digest but I no longer see us working together on future PRs. I simply do not have the patience to walk you through these things. Dealing with your over protectiveness is too much of a burden in reaching our shared goals.

Moving forward, I will be using [flask-cache-manifest](https://github.com/maxdup/flask-cache-manifest) instead. flask-cache-manifest doesn't feature the previously mentioned issues and gives me the peace of mind that it won't be a liability in my stack.

Trust me, I hate having to flex on you like this. My only comfort is knowing that in the end, open source always wins.
Have a good day.

nickjj commented:
Ok no problem, good luck.

> With no end in sight, I find myself having to abandon this PR. Do with it as you please.

If I do manually merge it, I'll for sure mention `@` mention you in the commit and also link back to your fork in case folks want to check it out.
